### PR TITLE
[TG-3694] Parse and store inner class information

### PR DESCRIPTION
**Pull request by jeannielynnmoulton**
This parses the InnerClasses attribute of the java bytecode with the intention of determining
- If an parsed class is an inner class
- The accessibility information for inner classes

If the parsed class is an inner class, then the original accessibility is overwritten by the information from the InnerClasses attribute.

Whether the class is an inner class is then stored in java_class_typet.

...

@thk123 @thomasspriggs Currently there is one test failing in test gen with the changes that will require some investigation, but I am on AL Friday and wanted to get some preliminary reviews so with some luck, this might be finished the sprint. Thanks.

@thk123 Could you please add an in progress label or equivalent. I have no privileges.

### CODE REVIEW

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp:1717]: This section should be moved into a new function/method, to avoid `rclass_attribute` getting too long.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: NIT: As this `pool_entry_lambda` is only used once, I would move it into the call to `.get_name`, rather than storing it in a variable.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Bike shedding: I would prefer `if(inner_class_info_index == 0) continue;` over indenting the remainder of the loop. I think opinions on this one are likely to vary.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I should probably add this exception to the coding standard, but I think `auto` is preferable for a lambda for two reasons:
1. By putting it into a `std::function` you allow the value to be null whereas the `auto` allows the compiler to pick a precise lambda type 
2. Probably doesn't help readability as the inputs and outputs are repeated on the rhs (`[this](u2 index) -> pool_entry &`)

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Nit: `const`

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: It is good that we have this check. However I think this code could be more robust, if we have the condition `number_of_bytes_to_be_read <= attribute_length` and a call to `skip_bytes` for the difference at the end of this block.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: You could use `parsed_class.is_inner_class = ` here instead of declaring an additional local variable and hence save a line of code inside the if block below.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Why does this work? Also possibly I think you might be able to check whether `outer_class_info_index == 0`? 

> If C is not a member of a class or an interface (that is, if C is a top-level class or interface (JLS §7.6) or a local class (JLS §14.3) or an anonymous class (JLS §15.9.5)), the value of the outer_class_info_index item must be zero.

Though that sounds like anonymous and local classes would be marked as not inner, which I guess is true?

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: You might consider pulling this out into a `read_innerclass_attribute` which would allow you to document when you'd expect to see this and an overview of the info we're getting out of it, as I can't get my head around when this attribute appears and what we find in it, despite reading the spec.

File: jbmc/src/java_bytecode/java_types.h
Comment (thk123) [inline jbmc/src/java_bytecode/java_types.h]: Return a bool (using `get_bool` as done in `get_final`) as this hides the messiness that everything is strings from the user of this function.

File: jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java]: Nit: missing new line

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: Nit: use `lookup_ref` if you are going to immediately dereference the return from `lookup` as it will produce a nicer error if the symbol doesn't exist.

File: jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java:1]: Could you add a non-public class to this file (in case that means there are two inner class attributes?) Unless that isn't true, in which case might not be worth it

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: I think this is wrong, I think the string will be something like "1" or "true" - which probably means if this passes CI these unit tests aren't being run.

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: Since you always load the same file, I'd probably do this at the top of scenario to reduce duplication

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp
Comment (thk123) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:1]: Missing from the `Makefile`.

**Review by thk123 — COMMENTED** [review]
Looks pretty good - couple of suggestions and I think your tests need tweaking.

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp
Comment (thk123) [reply jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: Indeed, it is the missing from Makefile as the CMake build fails:
https://travis-ci.org/diffblue/cbmc/jobs/397934975#L7792
`1 == is_inner_class`

File: jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java
Comment (thomasspriggs) [inline jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java:25]: Could you extend this test with some Inner Inner class examples, to show that things still work when the classes are nested an additional layer deep?

File: jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java
Comment (thomasspriggs) [reply jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java]: FYI: The clion/intellij settings can be tweaked to automatically add a new line to the end of files when it is missing. Should save a bunch of time on comments around this in your PRs.

File: jbmc/src/java_bytecode/java_types.h
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_types.h]: Could these methods be called `get_is_inner_class` and `set_is_inner_class` instead? The calls to the current method names superficially appear to be a call to a method on an outer class to get an inner class belonging to it.

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp
Comment (thomasspriggs) [inline jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp]: Can we also have some tests where an inner class is loaded first, instead of an outer class containing inner classes?

**Review by thomasspriggs — COMMENTED** [review]
Some nit picking. Couple of requests for extensions to the unit test.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Anonymous classes and local classes are both inner classes. If they are not being marked as such then we have a problem. I think the accessibility for these should always be private.

File: jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java
Comment (thomasspriggs) [reply jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java:25]: Could you also extend this with an anonymous class and a local class?

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Thinking about this, it may be advantageous to have a is_anonymous / is_local flag set at this point. We currently have an `is_anonymous` method in `state_typet`, which attempts to work this out based on regexing the name, which isn't the ideal approach.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: The spec is not great, agreed.

File: jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java
Comment (jeannielynnmoulton) [reply jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java]: Added and configured. 👍

File: jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java
Comment (jeannielynnmoulton) [reply jbmc/unit/java_bytecode/java_bytecode_parser/InnerClasses.java:1]: Added a non-public outer class 👍

File: jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp
Comment (jeannielynnmoulton) [reply jbmc/unit/java_bytecode/java_bytecode_parser/parse_java_attributes.cpp:1]: I am not sure what this means or how to add it. I will ask you when you are done with your current conversation.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: ~I've been asked to do worse. I am not sure which is clearer, but I've changed it~ 👍  ~No I haven't because the bytes need to be read anyway.~ No I am being stupid, changed :) Can you do double strikethrough?

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I see what you are saying, but I am not sure that it's best. If the bytes don't match, I don't think we can guarantee that anything makes sense because that would mean that the bytecode spec has changed basically, and so should probably just fail hard.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: NIT: These three lines give a clang-tidy warning in clion, due to the implicit cast from integer to bool. This could be fixed as follows - `bool is_protected = (inner_class_access_flags & ACC_PROTECTED) != 0;`

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: a129281e9ec21a24aeb1a228ca5096a5b0e7c867).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77913660

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: bccf9cc217d6944b919586733a445f750ebd598b).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77914476

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: The problems on test-gen were related to treating anonymous/local classes the same as inner classes. This has been updated by checking that outer_class_info_index  is 0 or not (it's 0 when it is outer, interface, local or anonymous). @thk123 The reason we can't just check outer_class_info_index is because the we only want to overwrite the accessibility information when the current parsed class (not the current parsed inner class information) is an inner class. As far as I can tell, there is no magic way to do this from the spec.

**Comment by jeannielynnmoulton** [comment]
@thk123 @thomasspriggs This is ready for re-review. Seems to have passed the compatibility check last night, which is what I would expect now that it is treating anonymous/local classes correctly and test-gen is still ignoring most of inner classes, so I'm reasonably confident that PR2014 will pass too. I doubt the test-gen changes one will make it into the sprint, but it's here anyway for reference.

Test-gen with changes: https://github.com/diffblue/test-gen/pull/2012

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Comment for why would be good here? My reading of the spec is it _can't_ be zero?

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Or a `read_bitflag` method but I wouldn't block this PR on it since this is a common approach used in this file.

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: c0c1029208a62433a3803b5aebc727337b22b31c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77954988

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Sorry if spotted this before, but having three booleans to represent a single value is error prone, I'd prefer an enum in the `classt`

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Nit: fewer words "Parses ~the bytes of~ the InnerClass attribute"

Comment (thk123) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Clarity: current parsed class or current parsed class file? (If the former, ignore me)

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Clarity: Inner classes of that class?

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thk123) [inline jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Probably stick this note by the code it relates to rather than here.

**Review by thk123 — APPROVED** [review]
All comments look to be addressed, couple of suggestions for making the comment on `rinner_classes_attribute` even better. I also think the enum for accessibility of `classt` would be a big win.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Could still do with a comment here.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Looking through the rest of this file, all the other instances of this have already been fixed to avoid the implicit cast.

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: This is a common approach in this file _and_ it effectively matches the java bytecode.

**Review by thomasspriggs — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 6ce7b1331c764d34c18f3402cfb27e0d58947c88).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77989278

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 9ba55e20aad17e94ac1fa6873681c20b3e3194c1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/77998632

**Comment by thk123** [comment]
On a subsequent PR you said you're just waiting on CI, but I think these two comments are outstanding: 
https://github.com/diffblue/cbmc/pull/2493/files#r199757047 (documenting the inner_class_index == 0 if)
https://github.com/diffblue/cbmc/pull/2493#discussion_r199468463 (using `!= 0` rather than implicit conversion to `bool`

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Removed

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (jeannielynnmoulton) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: Changed

**Comment by jeannielynnmoulton** [comment]
@thk123 I have removed the `if(inner_class_index == 0)` line, so that is no longer applicable. When someone says `NIT:` I (maybe wrongly) take that as optional, but I have changed it since I am pushing anyway.

**Comment by thk123** [comment]
Agree NIT means optional, but since there was another change (the docs, or as it turns out, removing the if) thought I'd draw your attention to it :slightly_smiling_face:. Will merge when passes CI.

**Comment by jeannielynnmoulton** [comment]
Waiting for CI to pass on bump: https://github.com/diffblue/test-gen/pull/2012

File: jbmc/src/java_bytecode/java_bytecode_parser.cpp
Comment (thomasspriggs) [reply jbmc/src/java_bytecode/java_bytecode_parser.cpp]: I considered it to be an optional "NIT" comment. I guess it would have been helpful if @thk123 had commented with something along the lines of "This is not a NIT, fix it."

**Review by allredj — COMMENTED** [review]
Passed Diffblue compatibility checks (cbmc commit: 5350133c8665f7f3f65190f6f51807e6296b3198).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/78105793

**Pull request merged**
